xarray/core/computation.py

```diff
@@ -0,0 +1,675 @@
+
+    input_data = [broadcast_compat_data(arg, broadcast_dims, core_dims)
+                  if isinstance(arg, Variable)
+                  else arg
+                  for arg, core_dims in zip(args, signature.input_core_dims)]
+
+    result_data = func(*input_data)
+
+    if signature.n_outputs > 1:
+        output = []
+        for dims, data in zip(output_dims, result_data):
+            output.append(Variable(dims, data))
+        return tuple(output)
+    else:
+        dims, = output_dims
+        return Variable(dims, result_data)
+
+
+def apply_array_ufunc(func, *args, **kwargs):
+    """apply_variable_ufunc(func, *args, dask_array='forbidden')
+    """
+    dask_array = kwargs.pop('dask_array', 'forbidden')
+    if kwargs:
+        raise TypeError('apply_array_ufunc() got unexpected keyword '
+                        'arguments: %s' % list(kwargs))
+
+    if any(isinstance(arg, dask_array_type) for arg in args):
+        # TODO: add a mode dask_array='auto' when dask.array gets a function for
+        # applying arbitrary gufuncs
+        if dask_array == 'forbidden':
+            raise ValueError('encountered dask array, but did not set '
+                             "dask_array='allowed'")
+        elif dask_array != 'allowed':
+            raise ValueError('unknown setting for dask array handling: %r'
+                             % dask_array)
+        # fall through
+    return func(*args)
+
+
+def apply_ufunc(func, *args, **kwargs):
+    """apply_ufunc(func, *args, signature=None, join='inner', new_coords=None,
+                   exclude_dims=frozenset(), dataset_fill_value=None,
+                   kwargs=None, dask_array='forbidden')
+
+    Apply a broadcasting function for unlabeled arrays to xarray objects.
+
+    The input arguments will be handled using xarray's standard rules for
+    labeled computation, including alignment, broadcasting and merging of
+    coordinates.
+
+    Parameters
+    ----------
+    func : callable
+        Function to call like ``func(*args, **kwargs)`` on unlabeled arrays.
+    *args : Dataset, DataArray, GroupBy, Variable, numpy/dask arrays or scalars
+        Mix of labeled and/or unlabeled arrays to which to apply the function.
+    signature : string or triply nested sequence, optional
+        Object indicating any core dimensions that should not be broadcast on
+        the input arguments, new dimensions in the output, and/or multiple
+        outputs. Two forms of signatures are accepted:
+        (a) A signature string of the form used by NumPy's generalized universal
+            functions [1], e.g., '(),(time)->()' indicating a function that
+            accepts two arguments and returns a single argument, on which all
+            dimensions should be broadcast except 'time' on the second argument.
+        (a) A triply nested sequence providing lists of core dimensions for each
+            variable, for both input and output, e.g., ([(), ('time',)], [()]).
```

Comment:
Yes, order of variables matters because that's how we match the signature up with `*args`.

To be clear, there is (currently) no special handling for matching ufunc arguments with `Dataset` variables by name.